### PR TITLE
test(ses): cover the allowNextEvalToBeUnsafe failure after stack overflow

### PR DESCRIPTION
**Pull request by naugtur**
## Description

Just a test covering an edge-case that's rare but known but often hard to recall.

### CODE REVIEW

File: packages/ses/test/make-safe-evaluator.test.js [hunk 1]
```diff
@@ -115,3 +115,31 @@ test('safeEvaluate - transforms - rewrite source', t => {
...
+    globalObject,
+    moduleLexicals,
+  });
+
+  let depth = 0;
+  function stackBuildUp() {
+    depth += 1;
+    evaluate(`1+1`);
+    // didn't fail yet? keep digging.
+    return stackBuildUp();
```
Comment (mhofman) [inline packages/ses/test/make-safe-evaluator.test.js:134]: I would recommend avoiding tail recursion here as that is an explicitly allowed optional optimization.

File: packages/ses/test/make-safe-evaluator.test.js [hunk 2]
```diff
@@ -115,3 +115,31 @@ test('safeEvaluate - transforms - rewrite source', t => {
...
+  }
+
+  t.throws(stackBuildUp, {
+    instanceOf: RangeError,
+    message: /Maximum call stack size exceeded/,
+  });
+  t.is(depth > 1, true, 'stack overflow occurred after many recursions');
+  t.throws(() => evaluate('1+1'), {
+    message: /a handler did not reset allowNextEvalToBeUnsafe \(a RangeError\)/,
+  });
```
Comment (mhofman) [inline packages/ses/test/make-safe-evaluator.test.js]: Interesting, I thought we couldn't actually trigger this case. We may have broken the stack overflow protection at some point.

**Review by mhofman — COMMENTED** [review]
I'm surprised you managed to build a test reproducing this. I'll try to look how we actually end up in this case.

File: packages/ses/src/eval-scope.js [hunk 3]
```diff
@@ -69,17 +75,35 @@ export const makeEvalScopeKit = () => {
     },
   });
 
+  const allowEvalProperties = freeze({
+    eval: {
+      value: props => {
+        const { eval: evalProp } = props;
+        delete evalScope.eval;
+        defineProperties(evalScope, { eval: evalProp });
+      },
```
Comment (naugtur) [inline packages/ses/src/eval-scope.js:84]: Took me a good 10 minutes of staring at this to figure out why it works.
I suggest we add a lengthy comment to explain it 😅 
Something like this maybe:

```suggestion
  // This conforms to the shape of evalScope properties, but instead of evaluating what arrives as first 
  // argument, it accepts a new shape to replace eval with. 
  // As a result, when `apply(evaluate, globalObject, [oneTimeEvalProperties]);` is called below,  
  // the same exact stack depth is reached and the same optimizations are applied to it before the  
  // FERAL_EVAL is put in place, making it impossible for a stack overflow error to occur in the process 
  // of getting eval for the main call to the evaluator if it didn't occur in revealing eval.
  const allowEvalProperties = freeze({
    eval: {
      value: props => {
        const { eval: evalProp } = props;
        delete evalScope.eval;
        defineProperties(evalScope, { eval: evalProp });
      },
```
